encodings/dbcs-codec.js

```diff
@@ -21,588 +18,603 @@ for (let i = 0; i < 0x100; i++) {
+    }
 
-        // Write the character to buffer, handling higher planes using surrogate pair.
-        if (uCode >= 0x10000) {
-            uCode -= 0x10000;
-            const uCodeLead = 0xd800 | (uCode >> 10);
-            newBuf[j++] = uCodeLead & 0xff;
-            newBuf[j++] = uCodeLead >> 8;
+    write(buf) {
+        const chars = this.backend.allocRawChars(buf.length),
+            prevBytes = this.prevBytes,
+            prevOffset = this.prevBytes.length;
+
+        let nodeIdx = this.nodeIdx,
+            seqStart = -this.prevBytes.length, // idx of the start of current parsed sequence.
+            charPos = 0;
+
+        for (let i = 0; i < buf.length; i++) {
+            const curByte = i >= 0 ? buf[i] : prevBytes[i + prevOffset];
+
+            // TODO: Check curByte is number 0 <= < 256
+
+            // Lookup in current trie node.
+            let uCode = this.decodeTables[nodeIdx][curByte];
+
+            if (uCode >= 0) {
+                // Normal character, just use it.
+            } else if (uCode === UNASSIGNED) {
+                // Unknown char.
+                // TODO: Callback with seq.
+                uCode = this.defaultCharUnicode.charCodeAt(0);
+                i = seqStart; // Skip one byte ('i' will be incremented by the for loop) and try to parse again.
+            } else if (uCode === GB18030_CODE) {
+                const b1 = i >= 3 ? buf[i - 3] : prevBytes[i - 3 + prevOffset];
+                const b2 = i >= 2 ? buf[i - 2] : prevBytes[i - 2 + prevOffset];
+                const b3 = i >= 1 ? buf[i - 1] : prevBytes[i - 1 + prevOffset];
+                const ptr =
+                    (b1 - 0x81) * 12600 + (b2 - 0x30) * 1260 + (b3 - 0x81) * 10 + (curByte - 0x30);
+                const idx = findIdx(this.gb18030.gbChars, ptr);
+                uCode = this.gb18030.uChars[idx] + ptr - this.gb18030.gbChars[idx];
+            } else if (uCode <= NODE_START) {
+                // Go to next trie node.
+                nodeIdx = NODE_START - uCode;
+                continue;
+            } else if (uCode <= SEQ_START) {
+                // Output a sequence of chars.
+                const seq = this.decodeTableSeq[SEQ_START - uCode];
+                for (let k = 0; k < seq.length - 1; k++) {
+                    uCode = seq[k];
+                    chars[charPos++] = uCode;
+                }
+                uCode = seq[seq.length - 1];
+            } else
+                throw new Error(
+                    `iconv-lite internal error: invalid decoding table value ${uCode} at ${nodeIdx}/${curByte}`
+                );
+
+            // Write the character to buffer, handling higher planes using surrogate pair.
+            if (uCode >= 0x10000) {
+                uCode -= 0x10000;
+                const uCodeLead = 0xd800 | (uCode >> 10);
+                chars[charPos++] = uCodeLead;
+
+                uCode = 0xdc00 | (uCode & 0x3ff);
+            }
+            chars[charPos++] = uCode;
```

Comment:
This is what I've hoped for - all decoders had to do this byte arithmetic themselves; now they just write 16-bit codes directly.